### PR TITLE
Restructure Quickstart docs to get from zero to hacking faster.

### PR DESCRIPTION
**Pull request by m-renaud**
I've made a few changes to the structure of the "Quickstart" section of the user guide so that someone new to Cabal can get to a Haskell package they can build and run sooner. Additionally, changes are made given the new Executable default for `cabal init` when non-interactive is used.

Changes include:
- Instructions start from an empty directory instead of "assuming we have a project directory and already have a Haskell module or two".
  - Rationale: If someone is reading the quickstart guide its unlikely they already know what a project directory looks like, and are unlikely have any modules already. They instead are probably looking for minimal instructions for how to get started with Haskell+Cabal.
- TL;DR; section at the top
  - Rationale: Sometimes you don't want to read a bunch of words, you just want a command to copy and paste into your terminal. This provides that.

This partially addresses #6194.

[ci skip]

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users have been recorded in the changelog.
* [x] The documentation has been updated, if necessary.
* [x] If the change is docs-only, `[ci skip]` is used to avoid triggering the build bots.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

File: Cabal/doc/developing-packages.rst [hunk 1]
```diff
@@ -1,83 +1,112 @@
...
+It is assumed that all the files that make up a package live under a common
+project root directory (apart from external dependencies). This simple
+example will have all the project files in one directory, but most packages
+will use one or more subdirectories.
+
+Cabal needs two extra files in the project's root directory:
+
+-  ``proglet.cabal``: contains package metadata and build information.
+
+-  ``Setup.hs``: contains a few standardized lines of code, but can be
```
Comment (vrom911) [inline Cabal/doc/developing-packages.rst]: This one is not mandatory if not using a custom set-up.

File: Cabal/doc/developing-packages.rst [hunk 2]
```diff
@@ -1,83 +1,112 @@
...
+   customized if necessary.
+
+
+Running ``cabal init`` interactively
```
Comment (vrom911) [inline Cabal/doc/developing-packages.rst]: I guess section above sections can go after this one because they are connected to `interactive` mode as well. For me this looks better structured:
* Using `cabal init` non-interactively (if this is the default one)
* Using `cabal init` interactively
* Building and running 
* ...

What do you think?

File: Cabal/doc/developing-packages.rst [hunk 3]
```diff
@@ -1,83 +1,112 @@
...
 
--  ``proglet.cabal``: containing package metadata and build information.
+Using ``cabal init``
+--------------------
 
--  ``Setup.hs``: usually containing a few standardized lines of code,
-   but can be customized if necessary.
+The ``cabal init`` command creates the necessary files for a Cabal
+package. It has an interactive mode which you can read more about below,
+but for now we're going to let it choose defaults for us.
```
[the remark this answers is not in the document]
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Done. Also, I ended up combining the two sections together, after moving them adjacent to each other as your suggested in a comment below it didn't make much sense for them to be separate.

File: Cabal/doc/developing-packages.rst [hunk 4]
```diff
@@ -1,83 +1,112 @@
...
-    ...
+    Generating Setup.hs...
+    Generating CHANGELOG.md...
+    Generating Main.hs...
+    Generating proglet.cabal...
 
-It also asks questions about various other bits of package metadata. For
-a package that you never intend to distribute to others, these fields
-can be left blank.
+Check out the files created using the ``ls`` command.
```
[the remark this answers is not in the document]
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Yup, definitely does, done :)

File: Cabal/doc/developing-packages.rst (hunk 1 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Added a note, I included it here for completeness because `cabal init` still creates it.

File: Cabal/doc/developing-packages.rst (hunk 2 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Much better, reworded the sections a bit as well. PTAL!

**Comment by phadej** [comment]
Note that audience have changed, from ”I cloned a Haskell repository and want to make a change” to ”I want to write a Haskell package”. 

IMHO former is more common, and latter is not something every user may even need. Given`cabal run`, and `cabal repl` there are simpler ways to start own experiments than creating own package. And creating own package is described (or should be) in ”development package”.

TL;DR don’t forget about the people, ”I cloned a package, what and how I can do next?”

**Comment by m-renaud** [comment]
Hey @phadej, thanks for the feedback!

> Note that audience have changed, from ”I cloned a Haskell repository and want to make a change” to ”I want to write a Haskell package”.

I don't think this is true, the audience was never for those cloning a Haskell repo. The previous wording said things like "Lets assume we have created a project directory and already have a Haskell module or two.", if you're at the point where you've found a Haskell repo you want to hack on you almost certainly know about `cabal build` and `cabal run`. Also, if they did just clone a repo there would already be a `.cabal` file and so the whole `cabal init` section wouldn't be relevant (which is the majority of the changes I've made in this PR). It was already targeted towards "I want to write a Haskell package" but had previous assumptions about the users knowledge which docs should never do (unless explicitly called out).

> IMHO former is more common, and latter is not something every user may even need.

I'm going to challenge this assumption, I wouldn't call myself a Haskell newbie by any means, and I've can `cabal init` to set up a directory for trying something small out orders of magnitude more times than I've cloned well established Haskell repos. It may be the case that for well versed Haskellers cloning an existing project and then hacking on that is the more common use case, and I admit that it is true that the current Haskell community consists of a high proportion of "experts", but the "Quickstart" docs should assume the opposite. 

> TL;DR don’t forget about the people, ”I cloned a package, what and how I can do next?”

For sure, this is the other common way to "get started" and doesn't really have any docs for it currently afaik. I still think the "Hey, I just learned about Haskell, I want to run through some tutorials, how do I create a file and set up some dependencies that I need?" should be the primary focus if we ever want to grow the Haskell community.

In general I think the overall docs need to be restructured to put these two "I'm just getting started"/quickstart sections much closer to the beginning. I have a bunch of ideas for improving the docs, wasn't sure the best way to solicit feedback before making a bunch of changes though, maybe open an RFC issue?

**Comment by tom-audm** [comment]
I respectfully don't understand the purpose of telling newcomers to `cabal init --non-interactive` in an empty directory:

  - If you're just doing it in an empty directory, you're not solving the case of "I just want a .cabal file so that I can `cabal v2-repl`," since there are no dependencies you're satisfying - you could simply `ghci`
  - If the user actually is creating a package, don't they want to specify things like license, author name, is this a library or executable, etc? Sensible defaults are given in interactive mode, too.

It would make most sense to me for the quickstart guide to either be the "what do I do with this cloned repo" that @phadej suggests, or to give the user hand-holding with `cabal init`.

**Comment by m-renaud** [comment]
Hey @tom-audm and @phadej, reviving this PR due to forced free time :P There are several changes in this PR and I'd like to get as many of them merged in as possible. I'm going to start by attempting to answer the questions you've raised:

> If you're just doing it in an empty directory...you could simply ghci

Neither using `ghci` nor `cabal v2-repl` create a `.cabal` file for you to add dependencies, so as soon as you want to use anything beyond what's in base (which is almost immediately because Haskell has a relatively small base library and relies on Hackage packages a lot), you need to figure out how to create a `.cabal` file anyways, so might as well start with a command that sets the environment up for you.

> If the user actually is creating a package, don't they want to specify things like license, author name, is this a library or executable, etc?

If you're a library author who's planning on uploading things to hackage than I think you're beyond the newcomer level and thus are already familiar with the tools. On the other hand, if you're a newcomer to Haskell all you want to know is "How do I get an environment up and running as soon as possible?" In general Haskell's docs throughout the ecosystem aren't the most newcomer friendly. If you look at other languages, for example the [Rust - Getting started](https://www.rust-lang.org/learn/get-started) page, it lays out going from zero to a functioning project with dependencies in a very concise way.

Reading through your comments again it sounds like these are the concerns you have:

- Previously this content only covered the ”I cloned a package, what and how I can do next?” (non-newcomer workflow), and now it only covers "Run `cabal init --non-interactive` to get a project set-up from scratch" (newcomer to the language workflow).
  - Proposal: Create two sub-sections: 1) Project from scratch, 2) Existing project, how do I hack on it?

- Why switch from interactive to non-interactive? We should let users know all of the options and configuration that is available.
  - Re-reading through the changes I made I do call out both interactive and non-interactive and describe what they do. Could you leave comments on the actual code changes and request specific changes so that it's actionable?

I think the major source of misalignment comes from me viewing the "Quickstart" as a "Haskell newcomer quickstart", and (my understanding of) your view as the "Quckstart" to "I want to develop a Haskell package to put on Hackage". The solution to this may be to create a separate "Getting Started" page in the Cabal docs (Proposal: "Section 0. Quickstart" or "Section 0. Getting started with Haskell and Cabal") which mirrors the "Rust - Getting started" page. If this sounds like a reasonable path forward I can put together a PR for comments.

Let me know what you think!

File: Cabal/doc/developing-packages.rst [hunk 5]
```diff
@@ -1,83 +1,119 @@
 Quickstart
 ==========
 
-Lets assume we have created a project directory and already have a
-Haskell module or two.
+Starting from scratch, we're going to walk you through creating a simple
+Haskell project.
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: One have to be very careful with words *project* and *package*.

This document doesn't create `cabal.project` file, so the *word* project should not be used at all. Alternatively one have to describe how project and package relate.

As this is page is targeted to beginners, we should not teach them anything we need then unteach later.

File: Cabal/doc/developing-packages.rst [hunk 6]
```diff
@@ -1,83 +1,119 @@
...
 
-Every project needs a name, we'll call this example "proglet".
+**TL;DR;** ``mkdir proglet && cd proglet && cabal init --non-interactive && cabal run``
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: isn't `--non-interactive` default now?

File: Cabal/doc/developing-packages.rst [hunk 7]
```diff
@@ -1,83 +1,119 @@
...
 
-We can create both files manually or we can use ``cabal init`` to create
-them for us.
+Using ``cabal init``
+--------------------
 
-Using "cabal init"
-------------------
+The ``cabal init`` command creates the necessary files for a Cabal package,
+it has both an ``--interactive`` (default) and ``--non-interactive``
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: haven't these reversed?

File: Cabal/doc/developing-packages.rst [hunk 8]
```diff
@@ -1,83 +1,119 @@
...
+can build and run our project.
+
+::
+
+   $ cabal build
+   Resolving dependencies...
+   ...
+   Linking /path/to/proglet ...
+
+   $ cabal run
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: I'd add the executable name. It's nice that `cabal run` works without it, but that is rare special case.

File: Cabal/doc/developing-packages.rst [hunk 9]
```diff
@@ -1,83 +1,119 @@
...
+   <Stuff about resolving dependencies, build profile, compliling, linking>
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: Just `...`

File: Cabal/doc/developing-packages.rst [hunk 10]
```diff
@@ -1,83 +1,119 @@
...
+   Hello, Haskell!
+
+Since we have an executable we can use ``cabal run`` which will build our
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: adopt to teach about `cabal run component-name`, that would be useful for tests too.

File: Cabal/doc/developing-packages.rst [hunk 11]
```diff
@@ -1,83 +1,119 @@
...
+executable (and re-build it if we've made any changes) and then run the
+binary.
+
+
+About the Cabal package structure
+---------------------------------
+
+It is assumed that all the files that make up a package live under a common
+project root directory (apart from external dependencies). This simple
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: again project vs. package. This is wrong. Just `common root directory` would be less ambiguous.

File: Cabal/doc/developing-packages.rst [hunk 12]
```diff
@@ -1,83 +1,119 @@
...
+example will have all the project files in one directory, but most packages
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: package files

File: Cabal/doc/developing-packages.rst [hunk 13]
```diff
@@ -1,83 +1,119 @@
...
+will use one or more subdirectories.
+
+Cabal needs two extra files in the project's root directory:
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: package's

File: Cabal/doc/developing-packages.rst [hunk 14]
```diff
@@ -1,83 +1,119 @@
...
+
+-  ``proglet.cabal``: contains package metadata and build information.
+
+-  ``Setup.hs``: contains a few standardized lines of code, but can be
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: In fact, when it's `build-type: Simple`, the contents **have to be standard**. Better would be for `cabal init` not to create `Setup.hs` at all.

File: Cabal/doc/developing-packages.rst [hunk 15]
```diff
@@ -166,23 +213,23 @@ package, so we must list it:
     library
       exposed-modules:     Proglet
       other-modules:
-      build-depends:       containers, base == 4.*
+      build-depends:       containers, base >=4.xx && <4.yy
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: Above was valid syntax, below isn't.

File: Cabal/doc/developing-packages.rst [hunk 16]
```diff
@@ -214,9 +262,14 @@ the package:
...
 
+As mentioned above, if you have an executable you can also run it with:
+
+.. code-block:: console
+
+   $ cabal run
+
 Assuming those two steps worked then you can also install the package:
 
 .. code-block:: console
```
Comment (phadej) [inline Cabal/doc/developing-packages.rst]: Does this work? It would be nice if it worked, but i'm 100% that installing local llbraries doesn't. (There should be a note, that it will work eventually, but doesn't now).

**Comment by phadej** [comment]
I'll change non-interactive `init` to not pick any license if not explicitly asked by a flag. You should adjust the text accordingly.

The side-effect of above would be that these test projects won't be as such uploadable to Hackage. 

> If you're a library author who's planning on uploading things to hackage than I think you're beyond the newcomer level and thus are already familiar with the tools. 

From my experience, that is not true. People want to upload their stuff to Hackage almost as soon as they get something working, sometimes even a bit before that.

File: Cabal/doc/developing-packages.rst (hunk 6 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Yup, updated, and also explicitly added `proglet` as the argument to `cabal run`.

File: Cabal/doc/developing-packages.rst (hunk 7 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Since I wrote this, yes. Move the (default) to after non-interactive.

File: Cabal/doc/developing-packages.rst (hunk 8 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Done.

File: Cabal/doc/developing-packages.rst (hunk 9 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Done.

File: Cabal/doc/developing-packages.rst (hunk 10 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Added some words, ptal!

File: Cabal/doc/developing-packages.rst (hunk 11 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Done.

File: Cabal/doc/developing-packages.rst (hunk 12 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Done.

File: Cabal/doc/developing-packages.rst (hunk 13 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Done.

File: Cabal/doc/developing-packages.rst (hunk 14 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Acknowledged. Currently `cabal init` still creates the `Setup.hs` file, if you can create an issue and assign it to me I can clean that up as a follow-up to this PR and update the docs.

File: Cabal/doc/developing-packages.rst (hunk 15 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Could you clarify? I'm not sure what you mean.

File: Cabal/doc/developing-packages.rst (hunk 16 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Hmm, I don't think so, but I'll need to investigate further. Going to call this out of scope of this change, can follow-up afterwards.

**Comment by m-renaud** [comment]
> I'll change non-interactive init to not pick any license if not explicitly asked by a flag. You should adjust the text accordingly.

SG. Alternatively I can submit this and you can make the doc change at the same time you change the behaviour (my preference to keep doc updates with code changes).

> From my experience, that is not true

Fair enough. I guess I was referring to folks who are just discovering haskell, it's unlikely they'll be uploading to hackage right away, but your point still stands. lmk if you have suggestions for wording tweaks around this.

File: Cabal/doc/developing-packages.rst (hunk 15 above)
Comment (phadej) [reply Cabal/doc/developing-packages.rst]: you can literally write `base ==4.*` in the cabal file, you cannot `>= 4.xx && <4.yy`. I don't like that we have code snippets which aren't "real".

IIRC there was an issue about the fact that we e.g. use `pkg-x.y.z` and sometimes that `pkg` is a variable, and sometimes it is literal. We don't have a solution to that (I'd use slated text for variables, but we have no means to format so), so better to stick to literal text when possible.

It is IMO lesser band if the bounds are different then what the current `cabal-install` executable would generate. I guess it depends on GHC generated, and that difference between text and what user sees ought to be explained.

TL;DR don't make user think which parts are "patterns" and which are literal.

File: Cabal/doc/developing-packages.rst (hunk 16 above)
Comment (phadej) [reply Cabal/doc/developing-packages.rst]: Add

```
.. todo:: this might or might not work with v2-install
```

I don't believe in follow-ups. Better to leave a sign.

Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Added todo.

File: Cabal/doc/developing-packages.rst (hunk 15 above)
Comment (m-renaud) [reply Cabal/doc/developing-packages.rst]: Ahh gotcha, that makes sense. Changed to `>=4.11 & <4.12` throughout page.

**Comment by m-renaud** [comment]
Alright, I think I've addressed all of your comments. Let me know if there's anything else, otherwise can you merge this into master?

**Comment by fgaz** [comment]
Why did this stall? Is it still up to date?

**Comment by m-renaud** [comment]
Just rebased but looks like files have changed locations since this was started.

**Comment by m-renaud** [comment]
Alright, fixed. The rebase blew away the comment anchors though unfortunately.

**Comment by m-renaud** [comment]
@phadej PTAL when you have a chance.

**Comment by Mikolaj** [comment]
Let me take over the review from @phadej. Unless @emilypi or @fgaz would like to review instead?

File: doc/developing-packages.rst [hunk 17]
```diff
@@ -367,7 +424,7 @@ Unit of distribution
 
 The Cabal package is the unit of distribution. What this means is that
 each Cabal package can be distributed on its own in source or binary
-form. Of course there may be dependencies between packages, but there is
+form. Of course there may dependencies between packages, but there is
```
Comment (Mikolaj) [inline doc/developing-packages.rst:427]: typo: the 'be' should remain here

**Review by Mikolaj — CHANGES_REQUESTED** [review]
I've reviewed rather shallowly, but apart of a few nitpicks it LGTM and based on the in-depth discussion by the other esteemed reviewers I'd love to see this merged ASAP (sorry for the prolonged process --- the kind souls helping with cabal are smeared rather thin).

File: doc/developing-packages.rst [hunk 18]
```diff
@@ -1,88 +1,124 @@
...
-can be left blank.
+    Generating Setup.hs...
+    Generating CHANGELOG.md...
+    Generating Main.hs...
+    Generating proglet.cabal...
 
-Finally, ``cabal init --interactive`` creates the initial ``proglet.cabal``
-and ``Setup.hs`` files, and depending on your choice of license, a
-``LICENSE`` file as well.
+Check out the files created using the ``ls`` command.
```
[the remark this answers is not in the document]
Comment (jneira) [reply doc/developing-packages.rst:72]: minor windows note: ls will not work with classic CMD in windows (it will do with powershell though)

**Comment by Mikolaj** [comment]
@m-renaud: do my nitpicks make sense? Let's get this merged.

**Comment by m-renaud** [comment]
Hey @Mikolaj! Yes, thank you for the review, sorry for the slow reply here; I'll try to get the changes out this week. It could definitely use some active voice rewording (read: more direct) throughout, but I can do that in a follow up.

File: doc/developing-packages.rst [hunk 19]
```diff
@@ -1,88 +1,124 @@
...
-After you make your selection (executable; library; or: library
-and executable) cabal asks us a number of questions starting with
-which version of the cabal specification to use, our package's name
-(for example, "proglet"), and our package's version.
+::
 
-It also asks questions about various other bits of package metadata. For
-a package that you never intend to distribute to others, these fields
-can be left blank.
+    Generating Setup.hs...
```
[the remark this answers is not in the document]
Comment (m-renaud) [reply doc/developing-packages.rst:67]: Removed! If folks are using an older version of cabal they can use the older version of the docs (that would still include info on Setup).

File: doc/developing-packages.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (m-renaud) [reply doc/developing-packages.rst:72]: Reworded. re. windows note, I don't have a windows machine (haven't for quite a while so don't know exactly what to put here). Can add windows notes in a follow-up.

File: doc/developing-packages.rst [hunk 20]
```diff
@@ -1,88 +1,124 @@
...
+our executable (and re-build it if we've made any changes) and then run the
+binary. The ``cabal run`` command works for any ``component-name`` (tests for
+example), not just the main executable.
+
+
+About the Cabal package structure
+---------------------------------
+
+It is assumed that all the files that make up a package live under a common
+root directory (apart from external dependencies). This simple example will
```
[the remark this answers is not in the document]
Comment (m-renaud) [reply doc/developing-packages.rst:107]: Done.

File: doc/developing-packages.rst [hunk 21]
```diff
@@ -1,88 +1,124 @@
...
+
+Cabal needs two extra files in the package's root directory:
+
+-  ``proglet.cabal``: contains package metadata and build information.
 
-    You may want to edit the .cabal file and add a Description field.
+-  ``Setup.hs``: contains a few standardized lines of code, but can be
+   customized if necessary.
+
+.. NOTE::
```
[the remark this answers is not in the document]
Comment (m-renaud) [reply doc/developing-packages.rst:118]: Removed Setup.hs references.

**Review by Mikolaj — APPROVED** [review]

**Review by emilypi — APPROVED** [review]

**Comment by Mikolaj** [comment]
Hi @m-renaud: I see there crept up conflicts since we last looked at this PR and I'm guilty of adding yet another conflict in #7526 (hurried up in order to backport to 3.6 in time; let's hope we can get this one to 3.6 as well).

Would you like to resolve the conflicts or should I do that?

**Comment by m-renaud** [comment]
Hey @Mikolaj! I'm on vacation until Monday but can resolve any merge conflicts next week, if that's an acceptable timeframe :)

**Comment by Mikolaj** [comment]
The rebased PR is at #7551. Whomever feels like it, please kindly review.

**Comment by Mikolaj** [comment]
This was merged as #7551. Thank you, @m-renaud and everybody!
